server/src/pages/shot/editor.js

```diff
@@ -104,61 +15,90 @@ exports.Editor = class Editor extends React.Component {
   }
 
   render() {
-    const color = this.isColorWhite(this.state.color);
-    const toolBar = this.cropToolBar || this.renderToolBar();
-    const display = this.loader || this.renderCanvas(color, toolBar);
+    const toolContent = this.renderSelectedTool();
+    const toolBar = this.renderToolBar();
+    const display = this.loader || this.renderCanvas(toolContent);
     return <div className="inverse-color-scheme full-height column-space">
       { toolBar }
       { display }
     </div>;
   }
 
-  renderCanvas(color, toolBar) {
+  renderCanvas(toolContent) {
     return <div className="main-container">
       <div
         className={`inverse-color-scheme canvas-container ${this.state.tool}`}
         id="canvas-container"
-        ref={(canvasContainer) => this.canvasContainer = canvasContainer}
         style={{ height: this.canvasHeight, width: this.canvasWidth }}>
         <canvas
           className="image-holder centered"
           id="image-holder"
           ref={(image) => { this.imageCanvas = image; }}
           height={this.canvasHeight * this.devicePixelRatio} width={this.canvasWidth * this.devicePixelRatio}
           style={{ height: this.canvasHeight, width: this.canvasWidth }}></canvas>
-        <canvas
-          className={`temp-highlighter centered ${color}`}
-          id="highlighter"
-          ref={(highlighter) => { this.highlighter = highlighter; }}
-          height={this.canvasHeight} width={this.canvasWidth}></canvas>
-        <div
-          className="crop-container centered"
-          ref={(cropContainer) => this.cropContainer = cropContainer}
-          style={{ height: this.canvasHeight, width: this.canvasWidth }}></div>
+        {toolContent}
       </div>
     </div>;
   }
 
+  renderSelectedTool() {
+    switch (this.state.tool) {
+      case "pen":
+        return <PenTool
+          ref={this.selectedTool}
+          color={this.state.color}
+          lineWidth={this.state.lineWidth}
+          baseCanvas={this.imageCanvas}
+          devicePixelRatio={this.devicePixelRatio}
+          updateImageCallback={this.onDrawingUpdate.bind(this)} />;
+      case "highlighter":
+        return <HighlighterTool
+          ref={this.selectedTool}
+          color={this.state.color}
+          lineWidth={this.state.lineWidth}
+          baseCanvas={this.imageCanvas}
+          devicePixelRatio={this.devicePixelRatio}
+          updateImageCallback={this.onDrawingUpdate.bind(this)} />;
+      case "cropTool":
+        return <CropTool
+          ref={this.selectedTool}
+          baseCanvas={this.imageCanvas}
+          devicePixelRatio={this.devicePixelRatio}
+          cancelCropHandler={this.onClickCancelCrop.bind(this)}
+          confirmCropHandler={this.onCropUpdate.bind(this)}
+          toolbarOverrideCallback={this.overrideToolbar.bind(this)} />;
+      default:
+        return null;
+    }
+  }
+
+  overrideToolbar() {
+    this.setState({overrideToolbar: this.state.tool});
```

Comment:
This changes the state just so that `renderToolBar()` is called _after_ the selected tool's been mounted in the previous render.